src/libmongoc/tests/mock_server/mock-server.c

```diff
@@ -488,43 +492,119 @@ mock_server_remove_autoresponder (mock_server_t *server, int id)
 
 
 static bool
-auto_ismaster (request_t *request, void *data)
+auto_hello_generate_legacy_response (request_t *request,
+                                     void *data,
+                                     bson_t *hello_response)
 {
    const char *response_json = (const char *) data;
    char *quotes_replaced;
-   bson_t response;
    bson_error_t error;
 
-   if (!request->is_command || strcasecmp (request->command_name, "ismaster")) {
-      return false;
-   }
-
    quotes_replaced = single_quotes_to_double (response_json);
 
-   if (!bson_init_from_json (&response, quotes_replaced, -1, &error)) {
+   if (!bson_init_from_json (hello_response, quotes_replaced, -1, &error)) {
       fprintf (stderr, "%s\n", error.message);
       fflush (stderr);
       abort ();
    }
 
+   bson_free (quotes_replaced);
+
+   return true;
+}
+
+static bool
+auto_hello (request_t *request, void *data)
+{
+   hello_callback_t *callback = (hello_callback_t *) data;
+   bson_t response;
+   bool is_hello;
+   bool is_legacy_hello;
+   char *response_json;
+   bson_iter_t iter;
+
+   if (!request->is_command) {
+      return false;
+   }
+
+   // Check whether we've got "hello" or legacy hello
+   is_hello = strcasecmp (request->command_name, "hello") == 0;
+   is_legacy_hello =
+      strcasecmp (request->command_name, HANDSHAKE_CMD_LEGACY_HELLO) == 0;
+
+   if (!is_hello && !is_legacy_hello) {
+      return false;
+   }
+
+   if (!callback->callback_func (request, callback->data, &response)) {
+      return false;
+   }
+
+   // Convert responses for legacy hello
+   if (bson_iter_init_find (&iter, &response, "isWritablePrimary")) {
+      BSON_APPEND_BOOL (
+         &response, HANDSHAKE_RESPONSE_LEGACY_HELLO, bson_iter_bool (&iter));
+   } else if (bson_iter_init_find (
+                 &iter, &response, HANDSHAKE_RESPONSE_LEGACY_HELLO)) {
+      BSON_APPEND_BOOL (&response, "isWritablePrimary", bson_iter_bool (&iter));
+   }
```

Comment:
This piece of code allows us to always use `nonWritablePrimary` in mocked responses, while retaining functionality for tests that might expect the legacy hello response.